### PR TITLE
feat(core): add single sign on authentication api

### PR DESCRIPTION
**Pull request by simeng-li**
## Summary

Implement the single sign on authentication api. Handles the following SSO specific interaction flow.
 - **sign-in with existing SSO identity** 
 - **sign-in and link SSO identity with same email**
 - **register with SSO identity**
 
This PR includes the following updates:

**Queries:**
- Create new query class for the `userSsoIdentities` table.  Including a custom query method `findUserSsoIdentityBySsoIdentityId`. By a given SSO `identityId`  and `issuers`. We should be able to uniquely identify a Logto user if exist.

**Refactors**
Refactor the `POST /interaction/sso/:connectorId/authorization-url` with following updates:

1. Update the input body guard, mark the `state` and `redirectUri` data optional, as SAML single-sign-on url generation does not request such inputs. The input body data will be guarded per connector instance. (OIDC or SAML)
2. Remove the old interaction event checking and guard logic. Intend to simplify the single sign on process. Unlike our existing interaction apis (will refactor later after the SSO implementation), no need to create a sign-in interaction by calling the `PUT /interaction` API ahead.  By calling this `POST /interaction/sso/:connectorId/authorization-url` we will automatically generate a `sign-in` interaction session with proper log and session storage. 
3. Extract the api logic in to a new created `single-sign-on.ts` util file as a util method `getSsoAuthorizationUrl`

**SSO classes**
Define a new required method for all the SSO connector classes `getIssuer()`. This method is used for fetch the unique identification of the connector's provider.  `Issuer` for OIDC connectors, IdP `entityId` for the SAML connectors.  This will be stored in the user's SSO identity data. Along with the SSO identityId, we could exclusively identify a user's SSO identity. 

**POST /interaction/sso/:connectorId/authentication**
This new API handles the final step of the Single Sign On process. The internal logic has been extracted into several utils methods:
- `getSsoAuthentication`: Using all the API provided body inputs and connector data in DB to get the full authentication info from SSO IdP using our SSO connector class.
- `handleSsoAuthentication` 
  - `signInWithSsoAuthentication`: directly sign-in if we find a matched user SSO identity 
  - `signInAndLinkWithSsoAuthentication`: Find a related email account if any. Automatically bind the new SSO identity to this matched account and sign-in.
  - `registerWithSsoAuthentication`: Create a new user account using the new SSO identity.

## Testing

It is kind of hard to implement a mock connector integration for now. Need to rely on the unit test for now. 

## Checklist

- [ ] `.changeset`
- [x] unit tests
- [ ] integration tests
- [ ] necessary TSDoc comments

### CODE REVIEW

File: packages/core/src/routes/interaction/utils/single-sign-on.ts [hunk 1]
```diff
@@ -0,0 +1,332 @@
...
+  // SignIn
+  if (userSsoIdentity) {
+    return signInWithSsoAuthentication(ctx, queries, {
+      connectorData,
+      userSsoIdentity,
+      ssoAuthentication,
+    });
+  }
+
+  const user = userInfo.email && (await usersQueries.findUserByEmail(userInfo.email));
```
Comment (wangsijie) [inline packages/core/src/routes/interaction/utils/single-sign-on.ts:148]: nit: just want to double check: our SSO implementation requires and only uses `email` as the identifier?

**Review by wangsijie — APPROVED** [review]

File: packages/core/src/queries/user-sso-identities.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  constructor(pool: CommonQueryMethods) {
+    super(pool, UserSsoIdentities);
+  }
+
+  async findUserSsoIdentityBySsoIdentityId(
+    issuer: string,
+    ssoIdentityId: string
+  ): Promise<Nullable<UserSsoIdentity>> {
+    return this.pool.maybeOne<UserSsoIdentity>(sql`
+      select user_id
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/core/src/queries/user-sso-identities.ts:26]: Good catch...

File: packages/core/src/sso/types/index.ts [hunk 3]
```diff
@@ -10,6 +10,7 @@ import { type JsonObject, type SsoConnector } from '@logto/schemas';
 export abstract class SingleSignOn {
   abstract data: SsoConnector;
   abstract getConfig: () => Promise<JsonObject>;
+  abstract getIssuer: () => string;
```
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/core/src/sso/types/index.ts:13]: yes, will rebase after your SAML PR is merged

File: packages/core/src/routes/interaction/utils/single-sign-on.ts (hunk 1 above)
Comment (simeng-li) [reply packages/core/src/routes/interaction/utils/single-sign-on.ts:148]: First of all, same as social, we uses SSO identities' userId as the SSOT to identify a user in Logto. 
In the product perspective, the Single Sign On flow was built based on the email identifier only. Only when user enters a email end with SSO registered domain will we trigger the Single Sign On flow. 
Specifically in this code block case, yes,   we will and will only identify the SSO identities' email and compare with our existing account, link them if a match is found.

File: packages/core/src/queries/user-sso-identities.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (simeng-li) [reply packages/core/src/queries/user-sso-identities.ts]: updated

File: packages/core/src/routes/sso-connector/index.ts [hunk 4]
```diff
@@ -29,9 +29,10 @@ export default function singleSignOnRoutes<T extends AuthedRouter>(...args: Rout
     router,
     {
       id: tenantId,
+      libraries: { ssoConnectors: ssoConnectorLibrary },
       queries: { ssoConnectors },
       libraries: {
-        ssoConnector: { getSsoConnectorById, getSsoConnectors },
+        ssoConnectors: { getSsoConnectorById, getSsoConnectors },
```
Comment (darcyYe) [inline packages/core/src/routes/sso-connector/index.ts:35]: can desctructure two `libraries` here?

**Review by darcyYe — APPROVED** [review]
LGTM

File: packages/core/src/routes/sso-connector/index.ts (hunk 4 above)
Comment (simeng-li) [reply packages/core/src/routes/sso-connector/index.ts:35]: should remove the above libraries.  Looks like a rebase issue

Comment (simeng-li) [reply packages/core/src/routes/sso-connector/index.ts:34]: updated

**Pull request merged**